src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs

```diff
@@ -51,19 +57,76 @@ public override HostServices HostServices
 
         public override Workspace Workspace => _workspace;
 
+        public override void Dispose()
+        {
+            ImmutableArray<IDisposable> disposableServices;
+            lock (_gate)
+            {
+                disposableServices = _ownedDisposableServices.ToImmutableArray();
+                _ownedDisposableServices.Clear();
+            }
+
+            // Take care to give all disposal parts a chance to dispose even if some parts throw exceptions.
+            List<Exception> exceptions = null;
+            foreach (var service in disposableServices)
+            {
+                try
+                {
+                    service.Dispose();
+                }
+                catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceledAndPropagate(ex))
+                {
+                    throw ExceptionUtilities.Unreachable;
+                }
+                catch (Exception ex)
+                {
+                    exceptions ??= new List<Exception>();
+                    exceptions.Add(ex);
+                }
+            }
+
+            if (exceptions is not null)
+            {
+                throw new AggregateException(CompilerExtensionsResources.Instantiated_parts_threw_exceptions_from_IDisposable_Dispose, exceptions);
+            }
+
+            base.Dispose();
+        }
+
         public override TWorkspaceService GetService<TWorkspaceService>()
         {
             if (TryGetService(typeof(TWorkspaceService), out var service))
             {
-                return (TWorkspaceService)service.Value;
+                // MEF workspace service instances created by a factory are not owned by the MEF catalog or disposed
+                // when the MEF catalog is disposed. Whenever we are potentially going to create an instance of a
+                // service provided by a factory, we need to check if the resulting service implements IDisposable. The
+                // specific conditions here are:
+                //
+                // * usesFactory: This is true when the workspace service is provided by a factory. Services provided
+                //   directly are owned by the MEF catalog so they do not need to be tracked by the workspace.
+                // * IsValueCreated: This will be false at least once prior to accessing the lazy value. Once the value
+                //   is known to be created, we no longer need to try adding it to _ownedDisposableServices, so we use a
+                //   lock-free fast path.
+                var checkAddDisposable = service.usesFactory && !service.lazyService.IsValueCreated;
+
+                var serviceInstance = (TWorkspaceService)service.lazyService.Value;
+                if (checkAddDisposable && serviceInstance is IDisposable disposable)
```

Comment:
this just seems so fragile.  i could easily imagine double disposes, or missing something for some random reason.  i really dont' like us adding complex Dispose logic in an already complex area.

Reply:
Currently the potential error here is a service can be requested after disposal, in which case the service will never be disposed. Changing this to throw ObjectDisposedException seemed too risky for this change.